### PR TITLE
Fix #587 Add a best effort cache option for list_libraries

### LINKED ISSUE #587: Add cache for list_libraries
#### Description of problem and/or code sample that reproduces the issue

List libraries is O(n) for number of libraries, so is slow for large numbers of databases and libraries, and over high latency links.  Provide a cache of the libraries in a top-level database to speed up initial connections.

### PR DESCRIPTION
**Pull request by shashank88**
list_libraries is a frequently called operation that can get
quite slow with a high amount of libraries. I noticed 4-5s+
for 7-8k libraries. Also the number of libraries is fairly
constant, so not much point in refetching the entire list every
time.

This adds a new collection in mongo which acts as a global cache
for all queries. The meta_db can be used for more accounting and
metadata storage in future, and this collection can be used for
caching other stuff as well.

It's best effort as if you don't have permissions or the data
does not exist in the cache it will just use the list_libraries.
The idea is to not require admin for reads and not expect the user
to have write access to a collection on writes.

### CODE REVIEW

File: arctic/arctic.py [hunk 1]
```diff
@@ -183,13 +188,17 @@ def __getstate__(self):
     def __setstate__(self, state):
         return Arctic.__init__(self, **state)
 
-    @mongo_retry
-    def list_libraries(self):
+    def list_libraries(self, newer_than_secs=120):
```
Comment (yschimke) [inline arctic/arctic.py]: It seems like there are a lot of dials for caching.  I think it's worth defaulting to the happy cache path.

e.g. This list_libraries seems useful to have a way to invalidate, but could be defaulted to -1.  That the cache value is always assumed to be ok.

It seems like there are a lot of ways e.g. mismatch between values where your cache will get ignored.

File: arctic/_cache.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        self._cachecol = None
+        try:
+            if cache_col not in self._cachedb.collection_names():
+                self._cachedb.create_collection(cache_col).create_index("date", expireAfterSeconds=cache_expiry)
+        except OperationFailure as op:
+            logging.debug("This is fine if you are not admin. The collection should already be created for you: %s", op)
+
+        self._cachecol = self._cachedb[cache_col]
+
+    def get(self, key, newer_than_secs=120):
```
Comment (yschimke) [inline arctic/_cache.py]: suggest -1 here be the default which means to always trust the cache

File: arctic/_config.py [hunk 3]
```diff
@@ -92,3 +92,11 @@ class FwPointersCfg(Enum):
 # ---------------------------
 # Configures the size of the workers pools used for async arctic requests
 ARCTIC_ASYNC_NWORKERS = os.environ.get('ARCTIC_ASYNC_NWORKERS', 4)
+
+
+# -------------------------------
+# Flag used for indicating caching levels. For now just for list_libraries.
+# -------------------------------
+ENABLE_CACHE = bool(os.environ.get('ENABLE_CACHE'))
```
Comment (yschimke) [inline arctic/_config.py]: Can this be defaulted to true?

File: arctic/arctic.py [hunk 4]
```diff
@@ -183,13 +188,17 @@ def __getstate__(self):
...
         """
         Returns
         -------
         list of Arctic library names
         """
+        _ = self._conn  # Ensure the connection exists and cache is initialized with it.
+        return self._list_libraries_cached(newer_than_secs) if ENABLE_CACHE else self._list_libraries()
```
Comment (yschimke) [inline arctic/arctic.py]: newer_than_secs == 0 probably means avoid checking the cache

File: arctic/arctic.py (hunk 1 above)
Comment (shashank88) [reply arctic/arctic.py]: Makes sense. Will default to the cached version with the default in config also to enable cache

**Pull request merged**